File: yoonju/Programmers/Level 3/입국심사.js
function solution(n, times) {
  const calTime = (value) => {
    return times.reduce((acc, cur) => acc + Math.floor(value / cur), 0);
  };

  var answer = Number(Infinity);

  times.sort((a, b) => a - b);

  const maxTime = times.at(-1);

  let [s, e] = [0, maxTime * n];

  while (s <= e) {
    const mid = Math.ceil((s + e) / 2);
    if (calTime(mid) >= n) {
      answer = Math.min(answer, mid);
      e = mid - 1;
    } else {
      s = mid + 1;
    }
  }
  return answer;
}

File: jieun/Programmers/입국심사.py
def solution(n, times):
    answer = max(times) * n  # 최대: 모두 시간이 오래 걸리는 심사를 받을 경우
    times.sort(reverse=True)

    def isPossible(sec):
        tmp_people = n
        for time in times:
            tmp_people -= sec // time  # sec//time: sec초 이내에 해결할 수 있는 최대 사람 수
            if tmp_people <= 0:
                return True
        return False

    def bs(left, right):
        nonlocal answer
        if left > right:
            return
        mid = (left + right) // 2
        if isPossible(mid):  # mid초 내에 해결 가능?
            answer = min(answer, mid)
            return bs(0, mid - 1)
        else:
            return bs(mid + 1, right)

    bs(0, max(times) * n)

    return answer
